### ISSUE TITLE
Sharded Dataset Has Long Delay Before First Batch (and caching=False)

### ISSUE THREAD

**Issue by cupdike** [webdataset/webdataset#267]
I have a sharded dataset for 1000 large images (675 MB each).  The shards were capped at 3 GB, and each shard contains 7 images (143 shards).

Using WebDataset with no caching enabled, when I benchmark how long it takes to iterate through the dataset using a batch size of 16, with 20 workers, I observe the following sequence of timing events:

- Metadata calls, HEAD, etc (takes 12 seconds)
- Tar downloads of all files (takes 93 seconds)
- No events occur for the next 613 seconds
- Batch 1 completes (613 seconds after tar download complete)
- Batch 2 completes (35 seconds after batch 1)
- The other batches start rapidly finishing
- Batch 20 (last) completes (20 seconds after batch 2)

Ideally, the batches would start processing sooner and with a regular cadence to keep the GPUs busy.  But I'm guessing the data download time is trivial compared to the time it takes to process the images into memory.  It seems like batches are all getting processed in unison rather than more serially.  Is there any way to improve this to get batches completed sooner and with a more even cadence to keep the GPU busy?

**Comment by cwerner** [comment]
I think I see this too.

 I use ˋpipe:gsutil catˋ on 254 shards of 0.7GB each. It takes forever before there is any GPU utilization… after like 5min I do get rapid progress with approx 80% gpu utilization. Next epoch, again a loooong wait with zero GPU activity?
How do I control caching if this could be the issue?

cheers

**Comment by cupdike** [comment]
So, I spent some time generating a chart of what I'm seeing.  I have not enabled caching programmatically and I verified the environment is clean of anything that might configure caching.  

The image below (sorry for the large size) shows elapsed seconds for the run on x, and file load duration seconds (log!) on y.  The symbols represent when the file load completed.  The labels along the bottom show which file (0-142) completed.  The labels along the top show when the batch completed.  

I guess I'm not totally sure what one should expect to see but I am seeing long delays between when files complete and when batches start getting emitted.  Also seeing long delays between when the last file completes and when the final batches are done.  There is also a lot of variability in how long it takes files to load (hence the need for the log scale).

Thoughts?

![image](https://github.com/webdataset/webdataset/assets/133214/7378eaef-5d33-4670-aa46-2ea0cdb5d3de)

**Comment by cwerner** [comment]
@cupdike Wondering if you did you fix this issue somehow?

**Comment by cupdike** [comment]
No.  Was surprised that the issues raised by that image didn't raise any eyebrows or merit a response.  Moved on...

**Comment by cwerner** [comment]
Same. May I ask what you are using instead? About to start a new project and wondering if I should switch? I need something with good performance in GCP…

**Comment by cupdike** [comment]
I had some other candidates to look at... Deep Lake and FFCV.  Never got around to it though.

**Comment by tmbdev** [comment]
I don't understand the bug report. You say that "Tar downloads of all files (takes 93 seconds)" and "Batch 1 completes (613 seconds after tar download complete)". But WebDataset doesn't download shards at all, it streams them and makes samples available as soon as they have been streamed, long before anything has been downloaded. That is the point of a streaming dataset library like WebDataset.

To see the actual latency and overhead of WebDataset, just run this very simple pipeline:

```
dataset = wds.WebDataset(url)
start = time.time()
data = next(iter(dataset))
finish = time.time()
print("time to first sample:", finish-start)
```

To see how many samples per second you get, use this:

```
dataset = wds.WebDataset(url)
start = time.time()
count = 0
for sample in dataset:
	count += 1
finish = time.time()
print("samples/second", count / (finish-start), "total", finish-start)
```

If you want to see how much overhead WebDataset adds, just run this:

```
start = time.time()
assert os.system("curl {url} > /dev/null") == 0
finish = time.time()
print("raw transfer total", finish-start)
```

If you are seeing large delays on startup, there are several possible reasons:

- large batch sizes that require many samples to download
- large shuffle buffers that take a long time to fill
- slow disk access or storage servers
- DataLoader related configurations

The most common cause of this is that you are using many workers (num_workers times number of GPUs) and that your dataset is stored on a single drive. In that case, I/O becomes your bottleneck. If you have 16 workers total and your dataset is stored on a single rotational drive, you are getting about 1-2 Mbytes/second I/O bandwidth for each worker, meaning it may take 600 seconds to load just a single image. That's a limitation of your hardware and your chosen configuration: a rotational drive can only deliver data at 20-200 Mbytes/second no matter what you do. The solution is to reduce num_workers to a smaller number, use SSDs, use caching on SSDs, and/or get more hardware. But this is only one of many possible configuration problems that lead to slow performance. Again, this has  nothing to do with WebDataset; you run into the same issue with any other data format.
